[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Bingjing-Dong/CSC448-AI/blob/main/Assignment%204.ipynb)

In [1]:
# Imports and pip installations (if needed)
import numpy as np
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely, not locally)
iris = datasets.load_iris()
X = iris.data #features (X) and label (y)
y = iris.target # 0 = Iris-Setosa, 1 = Iris-Versicolor, 2 = Iris-Virginica
z = iris.target_names

# Output the first 15 rows of the data
print('First 15 rows of data:')
df = pd.DataFrame(data= np.c_[X, y], columns= iris['feature_names'] + ['Species'])
df.head(15)

#for i in range(0,15):
#    print(i, X[i],z[0])

First 15 rows of data:


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    Species  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0  
5       0.0  
6       0.0  
7       0.0  
8       0.0  
9       0.0  
10      0.0  
11      0.0  
12      0.0  
13      0.0  
14      0.0

In [3]:
# Display a summary of the table information (number of datapoints, etc.)
print(iris['DESCR'])

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

## About the dataset
The Iris dataset contains four features: the length of sepal, the width of speal, the length of petal and the width of petal, all features in cm.  It has three species of Iris (Iris setosa, Iris versicolor and Iris virginica) and each of them has 50 samples. The feature X = iris.data and label y = iris.target(0 = Iris-Setosa, 1 = Iris-Versicolor, 2 = Iris-Virginica).

# Part 2: Split the dataset into train and test

In [4]:
# Take the dataset and split it into our features (X) and label (y)
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1, shuffle=True, random_state=58)
# If random_state change, the score below will be change as well.

# Part 3: Logistic Regression

In [5]:
# i. Use sklearn to train a LogisticRegression model on the training set
logistic_reg = LogisticRegression(max_iter=1000,random_state=42).fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
sample = np.array([[5.4, 3.7, 1.5, 0.2]])
print(f'Probabilites of each classes: {logistic_reg.predict_proba(sample)}')
# iii. Report on the score for Logistic regression model, what does the score measure?
score = logistic_reg.score(X_test, y_test)
print(f'Score of the Logistic regression model: {score}')
# iv. Extract the coefficents and intercepts for the boundary line(s)
print(f'Coefficents: {logistic_reg.coef_}')
print(f'Intercept: {logistic_reg.intercept_}')

Probabilites of each classes: [[9.75936614e-01 2.40633391e-02 4.65131142e-08]]
Score of the Logistic regression model: 0.8666666666666667
Coefficents: [[-0.43366109  0.86523868 -2.42414737 -1.04606788]
 [ 0.55920529 -0.49528279 -0.16643894 -1.01743846]
 [-0.1255442  -0.36995588  2.59058631  2.06350634]]
Intercept: [  9.83978208   2.41723642 -12.2570185 ]


The result indecate the sample datapoint in the features, 97.5% it belongs to class 0, 2.4% it belongs to class 1 and 0.0000046% it belongs to class 2. 

Score function returns the mean accuracy on the features_test and label_test. The score meaning 86.67% chance of classyfing correctly.

# Part 4: Support Vector Machine

In [6]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svc = SVC(C=1.0, probability=True).fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
print(f'Probabilites of each classes: {svc.predict_proba(sample)}')
# iii. Report on the score for the SVM, what does the score measure?
score = svc.score(X_test, y_test)
print(f'Score of the Support Vector Classifier: {score}')

Probabilites of each classes: [[0.97006762 0.01982866 0.01010373]]
Score of the Support Vector Classifier: 1.0


The result indecate the sample datapoint in the features, 97% it belongs to class 0, 1.98% it belongs to class 1 and 1.01% it belongs to class 2. 

Score function returns the mean accuracy on the features_test and label_test. The score meaning 100% chance of classyfing correctly.

# Part 5: Neural Network

In [7]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
MLP = MLPClassifier(max_iter=600,random_state=0).fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
print(f'Probabilites of each classes: {MLP.predict_proba(sample)}')
# iii. Report on the score for the Neural Network, what does the score measure?
score = MLP.score(X_test, y_test)
print(f'Score of the Neural Network: {score}')
# iv: Experiment with different options for the neural network, report on your best configuration 
#     (the highest score I was able to achieve was 0.8666)
MLP = MLPClassifier(max_iter=1000,random_state=42).fit(X_train, y_train)
print(f'New probabilites of each classes: {MLP.predict_proba(sample)}')
score = MLP.score(X_test, y_test)
print(f'Score of different options for the Neural Network: {score}')

Probabilites of each classes: [[9.99389045e-01 6.10954697e-04 6.84453272e-12]]
Score of the Neural Network: 0.9333333333333333
New probabilites of each classes: [[9.99580261e-01 4.19739134e-04 3.85682908e-15]]
Score of different options for the Neural Network: 0.9333333333333333


The result1 indecate the sample datapoint in the features, 99.9% it belongs to class 0, 0.061% it belongs to class 1 and 0.00000000068% it belongs to class 2. 

Score function1 returns the mean accuracy on the features_test and label_test. The score meaning 93.33% chance of classyfing correctly.

The result2 indecate the sample datapoint in the features, 99.9% it belongs to class 0, 0.041% it belongs to class 1 and 0.00000000000038% it belongs to class 2. 

Score function2 returns the mean accuracy on the features_test and label_test. The score meaning 93.33% chance of classyfing correctly.

# Part 6: K-Nearest Neighbors

In [9]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
# ii. For a sample datapoint, predict the probabilities for each possible class
KNN = KNeighborsClassifier().fit(X_train, y_train)
print(f'Probabilites of each classes: {KNN.predict_proba(sample)}')
# iii. Report on the score for kNN, what does the score measure?
score = KNN.score(X_test, y_test)
print(f'Score of the k-Neighbors Classifier: {score}')

Probabilites of each classes: [[1. 0. 0.]]
Score of the k-Neighbors Classifier: 0.8666666666666667


The result indecate the sample datapoint in the features, 100% it belongs to class 0, 0% it belongs to class 1 and 0% it belongs to class 2. 

Score function returns the mean accuracy on the features_test and label_test. The score meaning 86.67% chance of classyfing correctly.

# Part 7: Conclusions and takeaways

By looking the above work, all scores of different model is being good, the accuracy is above 86.67%. Support Vector Machine got the accuracy of 100%, which is the most fit for Iris dataset. Neural Network (MLP Classifier) also got accuracy of 93.33%, the result also shows us it is a good model.

One thing surprise me is that is if we change the parameters when creating model, the result score(accuracy) will be change slightly.